### PR TITLE
fix(langgraph-checkpoint-redis): scope vector search to the full namespace prefix

### PR DESCRIPTION
**Pull request by jackjin1997**
`RedisStore.search()` built two different namespace clauses from the same `namespacePrefix`:

| path | clause for `["tenant", "acme"]` |
| --- | --- |
| plain search | `@prefix:(tenant acme)` |
| vector search | `@prefix:tenant*` |

The vector path used `prefix.split(/[.-]/)[0]`, so every label after the first was dropped from the KNN filter, and the trailing `*` widened what was left. A vector search scoped to `["docs", "public"]` matched every namespace under `docs`, and `["tenant", "acme"]` also matched `tenants`. Every other namespace-clause site in the file (`get`, the existing-key check in `put`, `FilterBuilder.buildRedisSearchQuery`, and plain `search`) passes all tokens, so the vector path was the odd one out.

Both search paths now build the clause through one exported helper, `buildNamespacePrefixQuery`, so vector and plain search scope identically.

The helper also splits labels on the full set of punctuation RediSearch tokenizes on instead of only `.` and `-`. `prefix` is a TEXT field, so a label like `user:123` previously produced a term the indexer never created, and a label carrying `)` or `|` was interpolated into the clause verbatim. Now the clause can only contain terms the index could actually hold, so an injected payload degrades into extra AND terms, which narrows the match rather than widening it. A label made only of punctuation falls back to an escaped clause that matches nothing instead of `*`.

The same inline construction still sits at three other sites in `store.ts` (`get`, the existing-key check in `put`, `FilterBuilder.buildRedisSearchQuery`). Routing those through the helper touches read and write paths, so it is deliberately left out of this diff.

This does not change the segment-boundary contract discussed in #2721 for the Postgres and in-memory stores. `prefix` is a TEXT field, so token matching here stays order- and position-independent; making it exact needs an index-level change (the MongoDB store already stores a `namespacePath` array and matches element-wise). Happy to follow up on that separately if that direction is wanted.

## Tests

`src/tests/store.test.ts` (new) stubs the Redis client and asserts the query each search path actually issues, so it covers the fixed code path without a server: vector search keeps every label, no `@prefix:docs*` widening, plain search is unchanged, and an empty prefix still matches everything. `src/tests/utils.test.ts` covers the helper, including the tokenization and injection cases.

Reverting only `store.ts` fails the two vector-search assertions; reverting only the split regex in the helper fails 4 helper assertions. With both fixes in place the package suite is green.

```
pnpm --filter @langchain/langgraph-checkpoint-redis test    # 41 passed
pnpm --filter @langchain/langgraph-checkpoint-redis build   # ok (publint + attw clean)
oxfmt --check                                               # 441 files, correct format
oxlint libs/checkpoint-redis/src                            # 0 warnings, 0 errors
```

## AI Disclosure

This bug was identified through code review with AI assistance.

### CODE REVIEW

File: libs/checkpoint-redis/src/utils.ts [hunk 1]
```diff
@@ -21,6 +21,29 @@ export function escapeRediSearchTagValue(value: string): string {
...
+    .filter((token) => token.length > 0);
+
+  if (tokens.length === 0) {
+    return "*";
+  }
+
+  return `@prefix:(${tokens.join(" ")})`;
+}
+
 /**
```
[the remark this answers is not in the document]
Comment (jackjin1997) [reply libs/checkpoint-redis/src/utils.ts:68]: The injection surface was real — fixed in f22ca91, though not with the suggested escape.

`prefix` is indexed as a TEXT field (`"$.prefix": { type: "TEXT", AS: "prefix" }` in `SCHEMAS`), and `escapeRediSearchTagValue` is for TAG values inside `{...}`. Applying it here produces backslash-escaped literals, but the indexer tokenizes the stored prefix on punctuation, so a term like `user\:123` addresses something that was never indexed: the clause parses and then matches nothing. That would silently break every namespace label containing `:`, a space, or `@` — a worse failure than the one being reported, because it looks like data loss rather than a query error.

So `buildNamespacePrefixQuery` now splits labels on the same punctuation set RediSearch tokenizes on, which means the clause can only ever contain terms the indexer could have produced. `"acme) | @prefix:(victim"` becomes `@prefix:(tenant acme prefix victim)` — the payload degrades into extra AND terms, which narrows the match instead of widening it.

Two edge cases the new tests pin down:

- `[]` and `[""]` still mean "no prefix" and return `*`, unchanged.
- A label made *only* of punctuation indexes no term at all, so it now falls back to the escaped form (matches nothing). The obvious alternative — returning `*` when no tokens survive — would have widened the search to every namespace, which is the exact bug this PR is fixing.

Reverting just the split regex fails 4 of the new assertions, so they do cover the changed path.

One thing I deliberately left alone: the same unescaped `@prefix:(${tokens.join(" ")})` construction is still inline at three other sites in `store.ts` — `get`, the existing-key check in `put`, and `FilterBuilder.buildRedisSearchQuery`. Routing those through this helper is a two-line change, but it touches read and write paths, so I kept this diff to `search()`. Happy to fold them in here or open a follow-up, whichever is easier to review.

**Comment by jackjin1997** [comment]
Since #2721 now has five linked implementations (#2803, #2826, #2828, #2829, and #2830), I am closing this narrower patch to reduce parallel review. #2829 explicitly carries the Redis work forward with exact TAG semantics plus read and mutation coverage, and preserves credit for the shared query-helper approach. The focused regressions and security-hardening discussion remain available here; happy to revisit if maintainers prefer the smaller search-only change or the stacked proposal is re-scoped.

**Comment by hntrl** [comment]
Thank you @jackjin1997, will make sure you get properly attributed!
